src/services/codefixes/requireInTs.ts

```diff
@@ -5,25 +5,69 @@ namespace ts.codefix {
-            ? createImportDeclaration(/*decorators*/ undefined, /*modifiers*/ undefined, createImportClause(name, /*namedBindings*/ undefined), required)
-            : createImportEqualsDeclaration(/*decorators*/ undefined, /*modifiers*/ undefined, name, createExternalModuleReference(required)));
+    function doChange(changes: textChanges.ChangeTracker, sourceFile: SourceFile, info: Info) {
+        const { allowSyntheticDefaults, defaultImportName, namedImports, statement, required } = info;
+        changes.replaceNode(sourceFile, statement, defaultImportName && !allowSyntheticDefaults
+            ? createImportEqualsDeclaration(/*decorators*/ undefined, /*modifiers*/ undefined, defaultImportName, createExternalModuleReference(required))
+            : createImportDeclaration(/*decorators*/ undefined, /*modifiers*/ undefined, createImportClause(defaultImportName, namedImports), required));
     }
 
-    interface Info { readonly statement: VariableStatement; readonly name: Identifier; readonly required: StringLiteralLike; }
-    function getInfo(sourceFile: SourceFile, pos: number): Info {
+    interface Info {
+        readonly allowSyntheticDefaults: boolean;
+        readonly defaultImportName: Identifier | undefined;
+        readonly namedImports: NamedImports | undefined;
+        readonly statement: VariableStatement;
+        readonly required: StringLiteralLike;
+    }
+
+    function getInfo(sourceFile: SourceFile, program: Program, pos: number): Info | undefined {
         const { parent } = getTokenAtPosition(sourceFile, pos);
-        if (!isRequireCall(parent, /*checkArgumentIsStringLiteralLike*/ true)) throw Debug.failBadSyntaxKind(parent);
+        if (!isRequireCall(parent, /*checkArgumentIsStringLiteralLike*/ true)) {
+            throw Debug.failBadSyntaxKind(parent);
+        }
+
+        const allowSyntheticDefaults = getAllowSyntheticDefaultImports(program.getCompilerOptions());
         const decl = cast(parent.parent, isVariableDeclaration);
-        return { statement: cast(decl.parent.parent, isVariableStatement), name: cast(decl.name, isIdentifier), required: parent.arguments[0] };
+        const defaultImportName = tryCast(decl.name, isIdentifier);
+        const namedImports = allowSyntheticDefaults && isObjectBindingPattern(decl.name) ? tryCreateNamedImportsFromObjectBindingPattern(decl.name) : undefined;
+        if (!defaultImportName && !namedImports) {
+            return undefined;
+        }
+        return {
+            allowSyntheticDefaults,
+            defaultImportName,
+            namedImports,
+            statement: cast(decl.parent.parent, isVariableStatement),
+            required: first(parent.arguments)
+        };
+    }
+
+    function tryCreateNamedImportsFromObjectBindingPattern(node: ObjectBindingPattern): NamedImports | undefined {
+        const elements: ImportSpecifier[] = [];
+        for (const element of node.elements) {
+            if (isIdentifier(element.name)) {

```

Comment:
so this will drop any nested elements? I would prefer that test 3 show this instead of showing that the code fix isn't available (or perhaps it would be better to add a test):

`const { a: { b }, c } = require('a')` will produce `import { c } from 'a'`, right?

Reply:
Actually, I think the right behaviour would be to not offer the codefix if any element is not an identifier.

Reply:
> Actually, I think the right behaviour would be to not offer the codefix if any element is not an identifier.

Oke

Reply:
Sorry, I wasn't clear on what I thought the codefix should do. I think it should trigger on

```js
const a = require('a')
const { a } = require('a')
```

and it should not trigger on

```js
const { a: { b } } = require('ab')
```

However, the current state is fine too -- it fixes the bug and does so with a less-complicated arbitrary cutoff. Do you want to merge as-is or go back to supporting a single level of destructuring?

Reply:
I'll add a single level of destructuring